core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+                if (i === -1) {
+                  indexOfFirstDisplayedAction = 0;
+                }
+              };
+
+              expandActions();
+
+              // Index to know where to start highlighting actions.
+              let indexOfActionInFirstDisplayedBlock =
+                indexOfFirstDisplayedAction;
+
+              $scope.issueIsMultipleIncorrectSubmissions = false;
+              if (playthrough.issueType === (
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS)) {
+                $scope.issueIsMultipleIncorrectSubmissions = true;
+              }
+
+              /**
+               * Renders a table for MultipleIncorrectSubmissions issue for a
+               * more friendly UI.
+               */
+              $scope.renderIssueTable = function() {
+                var lars = LearnerActionRenderService;
+                var tableHtml =
+                  lars.renderFinalDisplayBlockForMISIssueHTML(
+                    playthrough.actions.slice(
+                      indexOfActionInFirstDisplayedBlock),
+                    indexOfActionInFirstDisplayedBlock + 1);
+                return tableHtml;
+              };
+
+              /**
+               * Returns the list of learner actions to currently be
+               * displayed in the playthrough modal.
+               * @returns {LearnerAction[]}
+               */
+              $scope.getActionsToRender = function() {
+                if ($scope.issueIsMultipleIncorrectSubmissions) {
+                  // This corresponds to the case where we display the table
+                  // for the issue.
+                  return playthrough.actions.slice(
+                    indexOfFirstDisplayedAction,
+                      indexOfActionInFirstDisplayedBlock);
+                }
+                return playthrough.actions.slice(indexOfFirstDisplayedAction);
+              };
+
+              /**
+               * Returns the index of the learner action.
+               * @param {LearnerAction} learnerAction.
+               * @returns {int}
+               */
+              $scope.getLearnerActionIndex = function(learnerAction) {
+                return playthrough.actions.indexOf(learnerAction) + 1;
+              };
+
+              /**
+               * Computes whether the learner action needs to be highlighted.
+               * @param {LearnerAction} learnerAction.
+               * @returns {boolean}
+               */
+              $scope.isActionHighlighted = function(action) {
+                return $scope.getLearnerActionIndex(action) > (
+                  indexOfActionInFirstDisplayedBlock);
+              };
+
+              /**
+               * Renders the HTML of the learner action.
+               * @param {LearnerAction} learnerAction.
+               * @param {int} actionIndex - The index of the learner action.
+               * @returns {string}
+               */
+              $scope.renderLearnerAction = function(
```

Comment:
This function name is showing up a lot throughout this PR. Is it necessary in all the places it is defined? I don't see it being used that much... so, please delete unused instances.

Reply:
@seanlip I was trying to stick to scope a bit. Keep the function visibility only to the necessary parts of the service.

Reply:
Oh yep, I understand that, but do you need all the different invocations of this function throughout the codebase? I'm not sure that all of them are being used.

Reply:
Discussed this, and it looks like all of them are actually being used